File: dynabuffers-python/dynabuffers/ast/ClassType.py
from dynabuffers.api.ISerializable import ISerializable, ByteBuffer
from dynabuffers.ast import FieldType
from dynabuffers.ast.datatype.OptionType import OptionType
from dynabuffers.ast.structural import ClassOptions


class ClassTypeOptions:
    def __init__(self, name, fields: [FieldType], options: ClassOptions):
        self.name = name
        self.fields = fields
        self.options = options


class ClassType(ISerializable):

    def __init__(self, options: ClassTypeOptions):
        self.options = options

    def size(self, value, registry):
        for field in self.options.fields:
            if field.options.defaultVal is None and field.options.name not in value:
                if not isinstance(field.options.datatype, OptionType):
                    raise ValueError("field '" + str(field.options.name) + "' is missing")
                else:
                    value[field.options.name] = None
            if field.options.defaultVal is None and value[field.options.name] is None:
                if not isinstance(field.options.datatype, OptionType):
                    raise ValueError("field '" + str(field.options.name) + "' is missing")
                else:
                    value[field.options.name] = None

        size = 0
        for field in self.options.fields:
            if field.options.name in value:
                size += field.size(value[field.options.name], registry)
            elif field.options.defaultVal is not None:
                size += field.size(field.options.defaultVal, registry)
        return size

    def serialize(self, value, buffer: ByteBuffer, registry):
        if self.options.options.is_deprecated():
            registry.addNotification("deprecated class " + self.options.name + " used")

        for field in self.options.fields:
            if not isinstance(field.options.datatype, OptionType):
                if field.options.defaultVal is None and field.options.name not in value:
                    raise ValueError("field '" + str(field.options.name) + "' is missing")
                if field.options.defaultVal is None and value[field.options.name] is None:
                    raise ValueError("field '" + str(field.options.name) + "' is missing")

        for element in list(
                filter(lambda x: (x.options.name in value) or x.options.defaultVal is not None, self.options.fields)):
            element.serialize(
                value[element.options.name] if element.options.name in value else element.options.defaultVal, buffer,
                registry)

    def deserialize(self, buffer, registry):
        map = {}
        for field in self.options.fields:
            map[field.options.name] = field.deserialize(buffer, registry)

        return map

File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/ClassType.kt
package dynabuffers.ast

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import dynabuffers.ast.datatype.OptionType
import dynabuffers.ast.structural.ClassOptions
import dynabuffers.exception.DynabuffersException
import java.nio.ByteBuffer
import java.util.*

data class ClassType(val options: ClassTypeOptions) : IType, ISerializable {

    override fun size(value: Any, registry: IRegistry): Int {
        val map = HashMap(value as Map<*, *>)

        for (field in options.fields) {
            if (map[field.options.name] == null && field.options.defaultVal == null) {
                if (field.options.dataType is OptionType)
                    map[field.options.name] = Optional.empty<Any>()
                else
                    throw DynabuffersException("field '${field.options.name}' is missing")
            }
        }

        return options.fields
                .filter { map.containsKey(it.options.name) || it.options.defaultVal != null }
                .map {
                    if (map.containsKey(it.options.name))
                        it.size(map[it.options.name]!!, registry)
                    else
                        it.size(it.options.defaultVal!!, registry)
                }.sum()
    }

    override fun serialize(value: Any, buffer: ByteBuffer, registry: IRegistry) {
        validate(registry)
        val map = HashMap(value as Map<*, *>)

        for (field in options.fields) {
            if (map[field.options.name] == null && field.options.defaultVal == null) {
                if (field.options.dataType is OptionType)
                    map[field.options.name] = Optional.empty<Any>()
                else
                    throw DynabuffersException("field '${field.options.name}' is missing")
            }
        }

        options.fields
                .filter { map.containsKey(it.options.name) || it.options.defaultVal != null }
                .forEach {
                    if (map.containsKey(it.options.name))
                        it.serialize(map[it.options.name]!!, buffer, registry)
                    else
                        it.serialize(it.options.defaultVal!!, buffer, registry)
                }
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any {
        val map = HashMap<String, Any>()
        options.fields.forEach { map.put(it.options.name, it.deserialize(buffer, registry)) }

        validate(registry)
        return map
    }

    private fun validate(registry: IRegistry) {
        if (options.options.isDeprecated()) {
            registry.addNotification("deprecated class ${options.name} used.")
        }
    }

    data class ClassTypeOptions(val name: String, val fields: List<FieldType>, val options: ClassOptions)

}
